man/php-fcgi.1.in
.\" PHP-FastCGI manualpage
.\"
.TH PHP-FCGI 1


.SH NAME
php-fcgi - start PHP as FastCGI daemon

.br
Warning, this tool is deprecated. Use php-fpm instead.


.SH SYNOPSIS
.B php-fcgi
[-c <configfile>] [-h] [-k] [-q] [-v]
.br
-c <configfile>: the configuration file to be used.
.br
-h: show help and exit.
.br
-k: kill running FastCGI servers.
.br
-q: don't print the results.
.br
-v: show version and exit.


.SH DESCRIPTION
The PHP-FastCGI utility can be used to start PHP as a FastCGI daemon. It uses php-cgi to do so.


.\" ==========[ Configuration ]================================================

.SH CONFIGURATION
The configuration is done via the configuration file ${CONFIG_DIR}/php-fcgi.conf.
.br

The following options are available:
.TP
.B MaxRequests = <number>
The number of maximum requests one fork will handle before dying and respawning.
.br
Default = 100, example: MaxRequests = 50

.TP
.B PIDfile = <path to PID-file>
This describes where php-fcgi should store the PID's of the FastCGI daemons.
.br
Default = ${PID_DIR}/php-fcgi.pid, example: PIDfile = /data/php-fcgi.pid

.TP
.B Setenv <key> = <value>
Sets an environment variable after clearing the entire environment and before starting the PHP FastCGI daemon.
.br
Example: Setenv somekey = somevalue

.TP
.B Server = <php-cgi executable> ; <forks> ; <binding> ; <uid>[:<gid's>][ ; <PHP configuration file>]
Defines how a FastCGI daemon should be executed. If a pipe-sign in <php-cgi executable> is used, the part before the pipe will be used as a chroot directory. Be carefull with using chrooted FastCGI servers in combination with UserWebsite and Alias (see hiawatha(1) for more information about these options). The userid and groupid 'root' are not allowed.
.br
Example: Server = /usr/bin/php5-cgi ; 3 ; 127.0.0.1:2005 ; www-data ; /etc/php5/cgi/php.ini
         Server = /usr/bin/php5-cgi ; 3 ; /tmp/hiawatha.sock ; 33:33


.SH SEE ALSO
The PHP-FastCGI utility is part of the Hiawatha webserver. See hiawatha(1) for more information about Hiawatha.


.SH AUTHOR
Hugo Leisink <hugo@hiawatha-webserver.org> - \fIhttp://www.hiawatha-webserver.org/\fP


ChangeLog
hiawatha (8.6) stable; urgency=low

  * PolarSSL updated to version 1.2. Added support for TLS 1.2 and
    secure renegotiation.
  * Added support for Server Name Indication.
  * MinSSLversion option added.
  * ServerRoot option removed.
  * Improved MacOS X package building script.
  * Marked php-fcgi as deprecated. Use php-fpm instead.
  * Small bugfixes and improvements.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 31 Oct 2012 19:10:32 +0100

hiawatha (8.5) stable; urgency=low

  * Improved Reverse Proxy.
  * Changed error message style.
  * Renamed Command Channel to Tomahawk.
  * Return 403 instead of 401 upon correct password for HTTP
    authentication but user not in right group.
  * Small improvements.
  * Bugfix: replaced select() with poll() to prevent crashes in case of
    large amount of simultaneous connections. Thanks to Peter Bex.

 -- Hugo Leisink <hugo@leisink.net>  Sun,  9 Sep 2012 11:39:12 +0200

hiawatha (8.4) stable; urgency=low

  * MaxServerLoad option added.
  * PolarSSL updated to version 1.1.4.
  * Small bugfixes and improvements.
  * Bugfix: invalid reverse proxy request when URL parameters are present.

 -- Hugo Leisink <hugo@leisink.net>  Thu,  7 Jun 2012 20:07:46 +0200

hiawatha (8.3.2) stable; urgency=high

  * Bugfix: memory leak in SSL library.

 -- Hugo Leisink <hugo@leisink.net>  Tue, 29 May 2012 18:02:59 +0200

hiawatha (8.3.1) stable; urgency=low

  * Improved security for reverse proxy (works with PreventSQLi, etc).

 -- Hugo Leisink <hugo@leisink.net>  Mon, 28 May 2012 21:50:31 +0200

hiawatha (8.3) stable; urgency=low

  * ReverseProxy option added.
  * PolarSSL updated to version 1.1.3.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 23 May 2012 18:11:56 +0200

hiawatha (8.2) stable; urgency=low

  * WebDAVapp option added. Enables support for WebDAV applications
    like ownCloud (http://owncloud.org/).
  * Removed support for the OPTIONS method.
  * AllowDotFiles option added.
  * Global forks setting in php-fcgi.conf moved to Server setting.
  * Small bugfixes and improvements.

 -- Hugo Leisink <hugo@leisink.net>  Tue,  1 May 2012 17:48:27 +0200

hiawatha (8.1) stable; urgency=low

  * BanOnInvalidURL option added.
  * PolarSSL updated to version 1.1.1.
  * Small improvements in Windows packaging script.
  * Bugfix: paths missing in default values and examples in manual pages.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 25 Feb 2012 19:02:41 +0100

hiawatha (8.0) stable; urgency=low

  * Replaced Autoconf with CMake. Many thanks to Sander Niemeijer.
  * Replaced OpenSSL with PolarSSL. Many thanks to Paul Bakker.
  * AllowedCiphers and DHparameters options removed.
  * Added IE7 to UrlToolkit's OldBrowser list, removed IE5.
  * MaxUrlLength option added, can return 414 Request-URI Too Long.
  * Changed default value of TriggerOnCGIstatus to 'no'.
  * Equalized format of logfiles.
  * Extra checks added to php-fcgi.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Fri, 27 Jan 2012 12:06:10 +0100

hiawatha (7.8.2) stable; urgency=high

  * Improved SQL injection detection.
  * Bugfix: memory leak in PreventSQLi routine.
  * Bugfix: potential server freeze with 100% CPU in CGI output caching.

 -- Hugo Leisink <hugo@leisink.net>  Fri, 18 Nov 2011 06:51:07 +0100

hiawatha (7.8.1) stable; urgency=low

  * Small bugfixes and improvements.
  * Bugfix: null byte in HTTP header of cached CGI content.

 -- Hugo Leisink <hugo@leisink.net>  Wed,  9 Nov 2011 17:21:52 +0100

hiawatha (7.8) stable; urgency=low

  * Control CGI output cache via X-Hiawatha-Cache and X-Hiawatha-Cache-Remove
    CGI headers. See the CGI OUTPUT CACHE section in the manual page.
  * BanOnWrongPassword now also triggers on wrong username.
  * Small improvements.
  * Bugfix: timeout issue with large POST requests on SSL connections.

 -- Hugo Leisink <hugo@leisink.net>  Mon, 31 Oct 2011 21:27:18 +0100

hiawatha (7.7) stable; urgency=low

  * First parameter of Alias can now contain subdirectories.
  * Improved stability for connections with SSL client authentication.
  * Bugfix: BanOnFlooding was broken.

 -- Hugo Leisink <hugo@leisink.net>  Tue,  4 Oct 2011 19:48:30 +0200

hiawatha (7.6) stable; urgency=low

  * PreventSQLi option rewritten.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 21 Aug 2011 08:06:21 +0200

hiawatha (7.5) stable; urgency=low

  * OldBrowser option added to URL toolkit.
  * Improved mimetype configuration.
  * Do-not-track HTTP header support.
  * Password file entries can now be created with Wigwam.
  * Small bugfixes and improvements.
  * Bugfix: sent one byte too few for Range -XX.
  * Bugfix: possible crash when using PreventSQLi.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 28 May 2011 15:39:13 +0200

hiawatha (7.4.1) stable; urgency=high

  * Bugfix: integer overflow in fetch_request() which could
    lead to a server crash.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 26 Feb 2011 10:32:24 +0100

hiawatha (7.4) stable; urgency=medium

  * Connections per IP added to RequestLimitMask.
  * NoExtensionAs made a per-host setting.
  * Small bugfixes and improvements.
  * Bugfix: usage of HideProxy caused Hiawatha to refuse new connections
    after ConnectionsTotal connections.
  * Bugfix: memory leak in XSLT module.

 -- Hugo Leisink <hugo@leisink.net>  Mon,  8 Nov 2010 20:58:54 +0100

hiawatha (7.3) stable; urgency=low

  * RequestLimitMask option added.
  * URL parameters for ErrorHandler.
  * Support for Haiku OS.
  * Small security bugfixes.

 -- Hugo Leisink <hugo@leisink.net>  Sun,  6 Jun 2010 23:18:37 +0200

hiawatha (7.2) stable; urgency=low

  * URL toolkit code restructured.
  * UseSSL option added to URL toolkit.
  * Digest HTTP authentication works with htdigest(1) created password files.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 21 Apr 2010 18:12:37 +0200

hiawatha (7.1) stable; urgency=low

  * Small bugfixes.
  * Bugfix: deny access and redirect result via toolkit subroutine.
  * Bugfix: broken flooding protection.


src/php-fcgi.c
/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License. For a copy,
 * see http://www.gnu.org/licenses/gpl-2.0.html.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 */

#include "config.h"
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <signal.h>
#include <grp.h>
#include <time.h>
#include <sys/wait.h>
#include "alternative.h"
#include "libstr.h"
#include "liblist.h"
#include "userconfig.h"

#define MAX_CONFIG_LINE KILOBYTE
#define DELIMITER ';'
#define TIMESTAMP_SIZE 40

typedef struct type_php_cgi {
	char     *executable;
	char     *chroot;
	char     *children;
	char     *binding;
	uid_t    uid;
	gid_t    gid;
	char     *config_file;
	t_groups groups;
	pid_t    pid;

	struct type_php_cgi *next;
} t_php_cgi;

typedef struct {
	char       *pidfile;
	char       *max_requests;
	t_php_cgi  *php_cgi;
	t_keyvalue *envir;
} t_config;

bool quiet = false;

/* Log a message
 */
void log_string(char *mesg, ...) {
	va_list args;
	FILE *fp;
	time_t t;
	struct tm *s;
	char timestamp[TIMESTAMP_SIZE];

	va_start(args, mesg);

	if (quiet == false) {
		vfprintf(stderr, mesg, args);
		fprintf(stderr, "\n");
	}

	if ((fp = fopen(LOG_DIR"/php-fcgi.log", "a")) != NULL) {
		time(&t);
		s = localtime(&t);
		timestamp[TIMESTAMP_SIZE - 1] = '\0';
		strftime(timestamp, TIMESTAMP_SIZE - 1, "%a %d %b %Y %T %z", s);

		fprintf(fp, "%s|", timestamp);
		vfprintf(fp, mesg, args);
		fprintf(fp, "\n");

		fclose(fp);
	}

	va_end(args);
}

/* Perform checks
 */
int check_file_access(char *file) {
	FILE *fp;

	if ((fp = fopen(file, "r")) == NULL) {
		fprintf(stderr, "Error: can't read %s.\n", file);
		return -1;
	}

	fclose(fp);

	return 0;
}

/* Start php-cgi as a FastCGI daemon
 */
int run_php_cgi(t_php_cgi *php_cgi) {
	FILE *fp;
	pid_t pid;

	if ((fp = fopen(php_cgi->executable, "r")) == NULL) {
		log_string("PHP binary %s not found", php_cgi->executable);
		return -1;
	}
	fclose(fp);

	switch (pid = fork()) {
		case -1:
			log_string("fork error");
			return -1;
		case 0:
			if (setsid() == -1) {
				exit(EXIT_FAILURE);
			}
			break;
		default:
			return pid;
	}

	setenv("PHP_FCGI_CHILDREN", php_cgi->children, 1);

	if (php_cgi->chroot != NULL) {
		if (chdir(php_cgi->chroot) == -1) {
			exit(EXIT_FAILURE);
		} else if (chroot(php_cgi->chroot) == -1) {
			exit(EXIT_FAILURE);
		}
	}

	do {
		if (setgroups(php_cgi->groups.number, php_cgi->groups.array) != -1) {
			if (setgid(php_cgi->gid) != -1) {
				if (setuid(php_cgi->uid) != -1) {
					break;
				}
			}
		}
		log_string("error while changing uid/gid");
		exit(EXIT_FAILURE);
	} while (false);

	/* Perform checks
	 */
	if (check_file_access("/etc/passwd") == -1) {
		exit(EXIT_FAILURE);
	}
	if (check_file_access("/dev/null") == -1) {
		exit(EXIT_FAILURE);
	}
	if (php_cgi->config_file != NULL) {
		if (check_file_access(php_cgi->config_file) == -1) {
			exit(EXIT_FAILURE);
		}
	}

	/* Redirect I/O
	 */
	if (close(STDIN_FILENO) == -1) {
		log_string("error closing stdin");
	} else if (open("/dev/null", O_RDONLY) == -1) {
		log_string("error redirecting stdin");
		exit(EXIT_FAILURE);
	}
	if (close(STDOUT_FILENO) == -1) {
		log_string("error closing stdout");
	} else if (open("/dev/null", O_WRONLY) == -1) {
		log_string("error redirecting stdout");
		exit(EXIT_FAILURE);
	}
	if (close(STDERR_FILENO) == -1) {
		log_string("error closing stderr");
	} else if (open("/dev/null", O_WRONLY) == -1) {
		quiet = true;
		log_string("error redirecting stderr");
		exit(EXIT_FAILURE);
	}

	if (php_cgi->config_file == NULL) {
		execlp(php_cgi->executable, php_cgi->executable, "-b", php_cgi->binding, (char*)NULL);
	} else {
		execlp(php_cgi->executable, php_cgi->executable, "-b", php_cgi->binding, "-c", php_cgi->config_file, (char*)NULL);
	}

	quiet = true;
	log_string("error while executing %s", php_cgi->executable);

	exit(EXIT_FAILURE);
}

/* Parse a configuration line
 */
int parse_line(t_config *config, char *line) {
	char *key, *value, *item, *group, *pipe;
	t_php_cgi *new;

	if (split_configline(line, &key, &value) != -1) {
		strlower(key);
		if (strcmp(key, "pidfile") == 0) {
			free(config->pidfile);
			if ((config->pidfile = strdup(value)) == NULL) {
				return -1;
			}
		} else if (strcmp(key, "maxrequests") == 0) {
			free(config->max_requests);
			if (str2int(value) < 1) {
				return -1;
			}
			if ((config->max_requests = strdup(value)) == NULL) {
				return -1;
			}
		} else if (strcmp(key, "setenv") == 0) {
			if (parse_keyvalue(value, &(config->envir), "=") == -1) {
				return -1;
			}
		} else if (strcmp(key, "server") == 0) {
			if ((new = (t_php_cgi*)malloc(sizeof(t_php_cgi))) == NULL) {
				return -1;
			}

			/* PHP executable
			 */
			if (split_string(value, &item, &value, DELIMITER) == -1) {
				return -1;
			}
			if ((pipe = strchr(item, '|')) != NULL) {
				*pipe = '\0';
				if ((new->chroot = strdup(item)) == NULL) {
					return -1;
				}
				*pipe = '/';
				item = pipe;
			} else {
				new->chroot = NULL;
			}
			if ((new->executable = strdup(item)) == NULL) {
				return -1;
			}

			/* Children
			 */
			if (split_string(value, &item, &value, DELIMITER) == -1) {
				return -1;
			}
			if (str2int(item) == -1) {
				return -1;
			}
			if ((new->children = strdup(item)) == NULL) {
				return -1;
			}

			/* Binding
			 */
			if (split_string(value, &item, &value, DELIMITER) == -1) {
				return -1;
			}
			if ((new->binding = strdup(item)) == NULL) {
				return -1;
			}

			/* UID, GID's
			 */
			split_string(value, &item, &value, DELIMITER);
			split_string(item, &item, &group, ':');
			if (parse_userid(item, &(new->uid)) != 1) {
				log_string("invalid user %s", item);
				return -1;
			}
			if (group != NULL) {
				if (parse_groups(group, &(new->gid), &(new->groups)) != 1) {
					log_string("invalid group %s", group);
					return -1;
				}
			} else {
				if (lookup_group_ids(new->uid, &(new->gid), &(new->groups)) != 1) {
					return -1;
				}
			}

			/* Configuration file
			 */
			if (value != NULL) {
				if ((new->config_file = strdup(value)) == NULL) {
					return -1;
				}
			} else {
				new->config_file = NULL;
			}

			new->next = config->php_cgi;
			config->php_cgi = new;
		} else {
			return -1;
		}
	}

	return 0;
}

/* Read the configuration file
 */
t_config *read_config(char *config_file) {
	t_config *config;
	FILE *fp;
	char line[MAX_CONFIG_LINE + 1], *data;
	int linenr = 0;

	if ((config = (t_config*)malloc(sizeof(t_config))) == NULL) {
		return NULL;
	}

	/* Default settings
	 */
	config->pidfile = strdup(PID_DIR"/php-fcgi.pid");
	config->max_requests = strdup("100");
	config->php_cgi = NULL;
	config->envir = NULL;

	if ((fp = fopen(config_file, "r")) == NULL) {
		perror(config_file);
		return NULL;
	}

	line[MAX_CONFIG_LINE] = '\0';
	while (fgets(line, MAX_CONFIG_LINE, fp) != NULL) {
		linenr++;
		data = uncomment(line);
		if (*data != '\0') {
			if (parse_line(config, data) == -1) {
				log_string("syntax error in %s on line %d", config_file, linenr);
				return NULL;
			}
		}
	}
	fclose(fp);

	return config;
}

/* Verify the configuration
 */
bool valid_config(t_config *config) {
	t_php_cgi *php_cgi;
	size_t len;

	php_cgi = config->php_cgi;
	while (php_cgi != NULL) {
		if ((len = strlen(php_cgi->executable)) >= 8) {
			if (strcmp(php_cgi->executable + len - 8, "php-fcgi") == 0) {
				log_string("%s is not a PHP binary", php_cgi->executable);
				return false;
			}
		}
		php_cgi = php_cgi->next;
	}

	return true;
}

/* Show help information
 */
void show_help(char *php_cgi) {
	printf("Usage: %s [options]\n", php_cgi);
	printf("Options: -c <configfile>: the configuration file to be used.\n");
	printf("         -h: show this information and exit.\n");
	printf("         -k: kill running FastCGI servers.\n");
	printf("         -q: don't print the results.\n");
	printf("         -v: show version and exit.\n");
}

/* Start FastCGI servers
 */
int start_fastcgi_servers(t_config *config) {
	FILE *fp;
	t_keyvalue *env;
	t_php_cgi *php_cgi;
	int started = 0, status;

	/* Start PHP FastCGI servers
	 */
	if (config->php_cgi == NULL) {
		fprintf(stderr, "No servers have been defined.\n");
		return 0;
	}

	clearenv();
	setenv("PHP_FCGI_MAX_REQUESTS", config->max_requests, 1);
	env = config->envir;
	while (env != NULL) {
		setenv(env->key, env->value, 1);
		env = env->next;
	}

	if ((fp = fopen(config->pidfile, "r")) != NULL) {
		fclose(fp);
		fprintf(stderr, "A PID file exists. Are FastCGI daemons already running?\n");
		return -1;
	}

	log_string("starting PHP FastCGI daemons");

	/* Start daemons
	 */
	php_cgi = config->php_cgi;
	while (php_cgi != NULL) {
		if ((php_cgi->pid = run_php_cgi(php_cgi)) != -1) {
			started++;
		}
		php_cgi = php_cgi->next;
	}
	if (started == 0) {
		return -1;
	}

	/* Check return codes
	 */
	sleep(1);
	php_cgi = config->php_cgi;
	while (php_cgi != NULL) {
		if (php_cgi->pid != -1) {
			if (waitpid(php_cgi->pid, &status, WNOHANG) > 0) {
				log_string("PHP FastCGI daemon exited with code %d", status);
				php_cgi->pid = -1;
			}
		}
		php_cgi = php_cgi->next;
	}

	/* Log PIDs
	 */
	if ((fp = fopen(config->pidfile, "w")) != NULL) {
		php_cgi = config->php_cgi;
		while (php_cgi != NULL) {
			if (php_cgi->pid != -1) {
				fprintf(fp, "%d\n", php_cgi->pid);
			}
			php_cgi = php_cgi->next;
		}
		fclose(fp);
	} else {
		fprintf(stderr, "write ");
		perror(config->pidfile);

		return -1;
	}

	return 0;
}

/* Stop running FastCGI servers
 */
int stop_fastcgi_servers(t_config *config) {
	FILE *fp;
	char line[11];
	size_t len;
	pid_t pid;

	if ((fp = fopen(config->pidfile, "r")) != NULL) {
		log_string("stopping PHP FastCGI daemons");

		line[10] = '\0';
		while (fgets(line, 10, fp) != NULL) {
			if ((len = strlen(line)) > 0) {
				if (line[len - 1] == '\n') {
					line[len - 1] = '\0';
				}
				if ((pid = (pid_t)str2int(line)) > 1) {
					if (kill(pid, SIGTERM) == -1) {
						fprintf(stderr, "kill -15 ");
						perror(line);
					}
				}
			}
		}

		fclose(fp);
		if (unlink(config->pidfile) == -1) {
			fprintf(stderr, "unlink ");
			perror(config->pidfile);

			return -1;
		}
	} else {
		fprintf(stderr, "read ");
		perror(config->pidfile);

		return -1;
	}

	return 0;
}

/* Main routine
 */
int main(int argc, char *argv[]) {
	t_config *config;
	char *config_file = CONFIG_DIR"/php-fcgi.conf";
	bool kill_servers = false;
	int i = 0;

	while (++i < argc) {
		if (strcmp(argv[i], "-c") == 0) {
			if (++i < argc) {
				config_file = argv[i];
			} else {
				fprintf(stderr, "Specify a configuration file.\n");
				return EXIT_FAILURE;
			}
		} else if (strcmp(argv[i], "-h") == 0) {
			show_help(argv[0]);
			return EXIT_SUCCESS;
		} else if (strcmp(argv[i], "-k") == 0) {
			kill_servers = true;
		} else if (strcmp(argv[i], "-q") == 0) {
			quiet = true;
		} else if (strcmp(argv[i], "-v") == 0) {
			printf("PHP-FastCGI v"VERSION". This tool is deprecated. Use php-fpm instead.\n");
			return EXIT_SUCCESS;
		} else {
			fprintf(stderr, "Unknown option. Use '-h' for help.\n");
			return EXIT_FAILURE;
		}
	}

	if ((config = read_config(config_file)) == NULL) {
		return EXIT_FAILURE;
	}
	if (valid_config(config) == false) {
		return EXIT_FAILURE;
	}

	if (kill_servers == false) {
		if (start_fastcgi_servers(config) == -1) {
			return EXIT_FAILURE;
		}
	} else {
		if (stop_fastcgi_servers(config) == -1) {
			return EXIT_FAILURE;
		}
	}

	return EXIT_SUCCESS;
}
